Dockerfile

```diff
@@ -16,27 +19,54 @@ RUN /src/scripts/docker-package.sh
 # Copy the config now so that we don't create another layer in the app image
 RUN cp /src/config.sample.json /src/webapp/config.json
 
-# App
-FROM nginx:alpine-slim
-
-COPY --from=builder /src/webapp /app
+FROM nginxinc/nginx-unprivileged:${DEBIAN_VERSION} AS nginx
 
 # Override default nginx config. Templates in `/etc/nginx/templates` are passed
 # through `envsubst` by the nginx docker image entry point.
 COPY /docker/nginx-templates/* /etc/nginx/templates/
 
-# Tell nginx to put its pidfile elsewhere, so it can run as non-root
-RUN sed -i -e 's,/var/run/nginx.pid,/tmp/nginx.pid,' /etc/nginx/nginx.conf
+USER root
+RUN rm -rf /usr/share/nginx/html
+COPY --from=builder /src/webapp /usr/share/nginx/html
 
-# nginx user must own the cache and etc directory to write cache and tweak the nginx config
-RUN chown -R nginx:0 /var/cache/nginx /etc/nginx
-RUN chmod -R g+w /var/cache/nginx /etc/nginx
+# Run a no-op action of nginx to run entrypoint scripts that may tweak config files
+USER nginx
+RUN ["nginx", "-t"]
+
+# App
+FROM gcr.io/distroless/base-nossl-debian${DEBIAN_VERSION_NUMERIC}:debug-nonroot
 
-RUN rm -rf /usr/share/nginx/html \
-  && ln -s /app /usr/share/nginx/html
+ARG TARGETARCH
+ARG LIBARCH=${TARGETARCH}
+ARG LIBARCH=${LIBARCH/amd64/x86_64}
+ARG LIBARCH=${LIBARCH/arm64/aarch_64}
+
+COPY --from=nginx \
+    /lib/${LIBARCH}-linux-gnu/ld-2.31.so \
+    /lib/${LIBARCH}-linux-gnu/libc-2.31.so \
+    /lib/${LIBARCH}-linux-gnu/libcrypt.so.1 \
+    /lib/${LIBARCH}-linux-gnu/libdl-2.31.so \
+    /lib/${LIBARCH}-linux-gnu/libpthread-2.31.so \
+    /lib/${LIBARCH}-linux-gnu/libz.so.1 \
+    /lib/${LIBARCH}-linux-gnu/
+
+COPY --from=nginx \
+    /usr/lib/${LIBARCH}-linux-gnu/libcrypto.so.1.1 \
+    /usr/lib/${LIBARCH}-linux-gnu/libpcre2-8.so.0 \
+    /usr/lib/${LIBARCH}-linux-gnu/libssl.so.1.1 \
+    /usr/lib/${LIBARCH}-linux-gnu/
+
+COPY --from=nginx /usr/sbin/nginx /usr/sbin/nginx
+COPY --from=nginx /usr/share/nginx/html /usr/share/nginx/html
+COPY --from=nginx /etc/nginx /etc/nginx
+COPY --from=nginx /var/log/nginx /var/log/nginx
```

Comment:
This looks pretty unmaintainable

Reply:
> What could help with stability is to lock down the Nginx image to a specific release. It also should be possible to write a script that checks which libraries are used by the base Nginx process & update this Dockerfile accordingly.

Couldn't the same script just set up a chroot in the originating image and copy that across to the app image?